File: matlab_files/b_diff_calc.m
function [] = b_diff_calc(b_diff_plots_folder,python_base_path,min_lat,max_lat,...
    min_lon,max_lon,min_depth,max_depth,figure_counter)
% This function calculates difference plots for the B coefficients
% Select runs for which they are different
runs = [25,26];

all_diff_files = {};
diff_paths = {};
count = 1;
for i=1:length(runs)
    diff_matrix_path = [python_base_path '\run_' num2str(runs(i)) ...
        '\difference_matrices_plots\viscosity\B_coeffs\'];
    diff_paths{i} = diff_matrix_path;
    dir_diff_files = dir([diff_matrix_path '\*.csv']);
    diff_files = {dir_diff_files.name};
    for j=1:length(diff_files)
        all_diff_files{count} = diff_files(j);
        count = count + 1;
    end
end

% Mesh data
resolution = 0.25;
latlim = [min_lat max_lat];
lonlim = [min_lon max_lon];
s = referenceSphere('Earth');
lat_lin = max_lat:-resolution:min_lat;
lon_lin = min_lon:resolution:max_lon;
[gridded_lon,gridded_lat] = meshgrid(lon_lin,lat_lin);
r_out = []; lon_plot_2 = []; lat_plot_2 = [];
depthrange = min_depth:1:max_depth;
for dd = depthrange
    lon_plot_2 = [lon_plot_2; gridded_lon(:)];
    lat_plot_2 = [lat_plot_2; gridded_lat(:)];
    temp = (s.Radius-dd*1e3)*ones(size(gridded_lon));
    r_out = [r_out; temp(:)];
end

% Find data matrix for the first run
matrix_1_flag = 0;
while matrix_1_flag == 0
    for i=1:length(all_diff_files)
        triplet_1 = [runs(1) min_depth max_depth];
        a = regexp(all_diff_files{i}, '\d+', 'match');
        found_values_1 = str2double(a{1});
        if isequal(triplet_1,found_values_1)
            matrix_1_flag = 1;
            file_to_read_1 = all_diff_files{i};
            break;
        end
    end
    if matrix_1_flag == 1
        break;
    else
        disp('No match for the selected values, the matrix does not exist. Exiting search.');
        break;
    end
end

% Find data matrix for the second run
matrix_2_flag = 0;
while matrix_2_flag == 0
    for i=1:length(all_diff_files)
        triplet_2 = [runs(2) min_depth max_depth];
        b = regexp(all_diff_files{i}, '\d+', 'match');
        found_values_2 = str2double(b{1});
        if isequal(triplet_2,found_values_2)
            matrix_2_flag = 1;
            file_to_read_2 = all_diff_files{i};
            break;
        end
    end
    if matrix_2_flag == 1
        break;
    else
        disp('No match for the selected values, the matrix does not exist. Exiting search.');
        break;
    end
end
min_depth = triplet_2(2);
max_depth = triplet_2(3);

% Again, filter the data, interpolate and plot it. Save the figure
% afterwards.
if matrix_1_flag == 1 && matrix_2_flag == 1
    % Extract the headers of both matrices, intersect and display the
    % possible common variables
    [values_1,headers_1,~] = xlsread([char(diff_paths{1}) char(file_to_read_1)]);
    [values_2,headers_2,~] = xlsread([char(diff_paths{2}) char(file_to_read_2)]);
    possible_variables_for_plotting = intersect(headers_1,headers_2, 'stable');
    possible_variables_for_plotting{end-2} = {}; 
    possible_variables_for_plotting{end-1} = {};
    possible_variables_for_plotting{end} = {};
    possible_variables_for_plotting = possible_variables_for_plotting(~cellfun('isempty',possible_variables_for_plotting));
    disp(['The possible variables to generate the difference plots between these'...
        ' two cycles are:\n']);
    disp(possible_variables_for_plotting);
    diff_variables_to_plot = possible_variables_for_plotting;
    filtered_lat = values_1(:,end-2);
    filtered_lon = values_1(:,end-1);
    depth_out = values_1(:,end);
    filtered_R = s.Radius - 1e3*depth_out;
    for i=1:length(diff_variables_to_plot)
        diff_variable_1 = values_1(:,strcmp(headers_1,diff_variables_to_plot{i}));
        diff_variable_2 = values_2(:,strcmp(headers_2,diff_variables_to_plot{i}));
        diff = diff_variable_2-diff_variable_1;
        [x_in,y_in,z_in]=sph2cart(deg2rad(filtered_lon),deg2rad(filtered_lat),filtered_R);
        [x_out,y_out,z_out]=sph2cart(deg2rad(lon_plot_2),deg2rad(lat_plot_2),r_out);
        plot_variable_out = griddata(x_in,y_in,z_in,diff,x_out,y_out,z_out,'nearest');
        
        figure(figure_counter)
        colormap summer;
        load coastlines;
        [Z, refvec] = geoloc2grid(lat_plot_2,lon_plot_2,plot_variable_out,resolution);
        ax = axesm('stereo','MapLatLimit',latlim,'MapLonLimit',lonlim,'Grid','on','MeridianLabel','on','ParallelLabel','on');
        set(ax,'Visible','off');
        cmap = colormap('summer');
        alpha 0.7;
        colormap(cmap);
        set(findall(gca, 'type', 'text'), 'visible', 'on')
        geoshow(Z, refvec, 'DisplayType', 'texture');
        plotm(coastlat, coastlon, 'color', rgb('OrangeRed'));
        caxis('auto'); % was [0 1e6]
        h = colorbar('v'); %set(h, 'ylim', [0 1e6]);
        if i == 1
            set(get(h,'ylabel'),'string','B_{diff}')
            title('Map of differences for B_{diff}');
        else
            set(get(h,'ylabel'),'string','B_{disl}')
            title('Map of differences for B_{disl}');
        end
        set(findall(gca, 'type', 'text'), 'visible', 'on')
        grid on;
        % Save the figures based on whether we are working with
        % stresses or deflections
        saveas(gcf,[b_diff_plots_folder 'diff_plot_' possible_variables_for_plotting{i} '_'...
            'depth_range_[' num2str(min_depth) '-' num2str(max_depth) ']_km.png']);
        figure_counter = figure_counter + 1;
    end
end
end

File: python_files_python_3/b_diff_calc.py
import os
import pdb
import numpy as np
import cartopy.crs as chart
import matplotlib.pyplot as plt
import scipy.interpolate as interp
import pandas as pd
import re
from matplotlib import cm
from plot_scale_extremes import *
from geo2cart import *
import math


def b_diff_calc(b_diff_plots_folder, python_base_path, min_lat, max_lat, min_lon, max_lon, min_depth,
                    max_depth, run_vec):

    b_runs = [run_vec[0], run_vec[1]]
    diff_files = []
    for i in range(len(b_runs)):
        diff_matrix_path = os.path.join(python_base_path, 'run_' + str(b_runs[i]),
                                  'difference_matrices_plots\\viscosity\\B_coeffs')
        dir_diff_files = os.path.join(diff_matrix_path, '*.csv')
        for j in range(len(dir_diff_files)):
            file_name = dir_diff_files[i].split('\\')
            diff_files.append(file_name[-1])

    resolution = 0.25
    earth_radius = 6371000
    lat_lin = np.arange(max_lat, min_lat, -resolution)
    lon_lin = np.arange(min_lon, max_lon, resolution)
    [gridded_lon, gridded_lat] = np.meshgrid(lon_lin, lat_lin)
    r_out = np.array([])
    lon_plot = np.array([])
    lat_plot = np.array([])
    depthrange = np.arange(min_depth, max_depth, 1)
    for dd in depthrange:
        lon_plot = np.vstack((lon_plot, gridded_lon)) if lon_plot.size else gridded_lon
        lat_plot = np.vstack((lat_plot, gridded_lat)) if lat_plot.size else gridded_lat
        temp = (earth_radius - dd * 1e3) * np.ones((len(gridded_lon), len(gridded_lon[0])))
        r_out = np.vstack([r_out, temp]) if r_out.size else temp

    matrix_1_flag = 0
    found_values_1 = []
    quintuplet_1 = np.arr([])
    while matrix_1_flag == 0:
        quintuplet_1 = input('Enter the iteration, step and cycle number, and the minimum and maximum depth of the first '
                                'matrix used to plot the difference with respect to another moment in time as a vector of '
                                'square brackets of five values: \n')
        for i in range(len(diff_files)):
            found_values_1 = re.findall(r"\((\d+(?:,\d+)*)\)", quintuplet_1)
            if (quintuplet_1 == found_values_1).all():
                matrix_1_flag = 1
                file_to_read_1 = diff_files[i]
                iteration_1 = quintuplet_1[0]
                step_1 = quintuplet_1[1]
                cycle_1 = quintuplet_1[2]
                break
        if matrix_1_flag == 1:
            break
        else:
            print('No match for the selected values, the matrix does not exist. Exiting search.')
            break

    found_values_2 = []
    matrix_2_flag = 0
    quintuplet_2 = np.arr([])
    while matrix_2_flag == 0:
        quintuplet_2 = input('Enter the iteration, step and cycle number, and the minimum and maximum depth of the first '
                                'matrix used to plot the difference with respect to another moment in time as a vector of '
                                'square brackets of five values: \n')
        for i in range(len(diff_files)):
            found_values_2 = re.findall(r"\((\d+(?:,\d+)*)\)", quintuplet_2)
            if (quintuplet_2 == found_values_2).all():
                matrix_1_flag = 1
                file_to_read_2 = diff_files[i]
                iteration_2 = quintuplet_2[0]
                step_2 = quintuplet_2[1]
                cycle_2 = quintuplet_2[2]
                break
        if matrix_2_flag == 1:
            break
        else:
            print('No match for the selected values, the matrix does not exist. Exiting search.')
            break

    if matrix_1_flag == 1 and matrix_2_flag == 1:
        headers_1 = pd.read_csv(os.path.join(diff_matrix_path, file_to_read_1), nrows = 1)
        headers_2 = pd.read_csv(os.path.join(diff_matrix_path, file_to_read_2), nrows = 1)
        values_1 = pd.read_csv(os.path.join(diff_matrix_path, file_to_read_1), skiprows = 1)
        values_2 = pd.read_csv(os.path.join(diff_matrix_path, file_to_read_2), skiprows = 1)
        possible_variables_for_plotting = headers_1.intersection(headers_2, 'stable')
        possible_variables_for_plotting[-3] = []
        possible_variables_for_plotting[-2] = []
        possible_variables_for_plotting[-1] = []
        print('The possible variables to generate the difference plots between these two cycles are: \n')
        print(possible_variables_for_plotting)
        diff_variables_to_plot = possible_variables_for_plotting
        filtered_lat = values_1[:, -3]
        filtered_lon = values_1[:, -2]
        depth_out = values_1[:, -1]
        filtered_r = earth_radius - 1e3 * depth_out
        for i in range(len(diff_variables_to_plot)):
            diff_variable_1 = values_1[:, (headers_1 == diff_variables_to_plot[i]).all()]
            diff_variable_2 = values_2[:, (headers_2 == diff_variables_to_plot[i]).all()]
            plot_variable = diff_variable_2 - diff_variable_1
            [x_in, y_in, z_in] = geo2cart(filtered_lat, filtered_lon, filtered_r)
            [x_out, y_out, z_out] = geo2cart(np.deg2rad(lon_plot), np.deg2rad(lat_plot), r_out)
            plot_variable_out = interp.griddata(x_in, y_in, z_in, plot_variable, x_out, y_out, z_out, 'nearest')

            # Open image
            diff_figure = plt.figure()
            # Geographic map
            ax = plt.axes(projection=chart.Orthographic(0, -90))
            ax.coastlines(resolution='50m', color='orange', linewidth=1)
            # Add surface
            ax.plot_surface(lon_plot, lat_plot, plot_variable_out, cmap=cm.summer, linewidth=0, antialiased=False)
            # Colorbar settings
            scale = plt.colorbar(diff_figure)
            diff_figure.clim('auto')
            scale = plt.colorbar(diff_figure)
            if i == 0:
                scale.set_label(diff_variables_to_plot[i])
            else:
                scale.set_label(diff_variables_to_plot[i])
            plt.savefig(os.path.join(b_diff_plots_folder, 'diff_plot_' + possible_variables_for_plotting[i] + '_' +
                                     '_depth_range_[' + str(min_depth) + '-' + str(max_depth) + ']_km.png'))

